orchestrator/careplancontributor/applaunch/zorgplatform/service.go

```diff
@@ -165,12 +168,141 @@ type Service struct {
 	zorgplatformCert       *x509.Certificate
 	profile                profile.Provider
 	secureTokenService     SecureTokenService
+	cpsTransport           http.RoundTripper
 }
 
 func (s *Service) RegisterHandlers(mux *http.ServeMux) {
 	mux.HandleFunc("POST "+appLaunchUrl, s.handleLaunch)
 }
 
+func (s *Service) BgzFhirProxy() coolfhir.HttpProxy {
+	targetFhirBaseUrl, _ := url.Parse(s.config.ApiUrl)
+
+	proxyBasePath := "/cpc/bgz/fhir"
+
+	rewriteUrl, _ := url.Parse(s.baseURL + proxyBasePath)
+
+	return coolfhir.NewProxy("App->EHR (BGZ)", log.Logger, targetFhirBaseUrl, proxyBasePath, rewriteUrl, stsAccessTokenRoundTripper{
+		transport:  s.zorgplatformHttpClient.Transport,
+		zpfService: s,
+	}, true)
+}
+
+var _ http.RoundTripper = &stsAccessTokenRoundTripper{}
+
+type stsAccessTokenRoundTripper struct {
+	transport  http.RoundTripper
+	zpfService *Service
+}
+
+func (s stsAccessTokenRoundTripper) RoundTrip(httpRequest *http.Request) (*http.Response, error) {
+	// Do something to request the access token
+	newHttpRequest := httpRequest.Clone(httpRequest.Context())
+
+	//TODO: Check if we can update the CarePlan.basedOn to point to the service request
+	carePlanReference := httpRequest.Header.Get("X-Scp-Context")
+	if carePlanReference == "" {
+		return nil, fmt.Errorf("missing X-Scp-Context header")
+	}
+
+	log.Debug().Ctx(httpRequest.Context()).Msgf("Found SCP context: %s", carePlanReference)
+	//TODO: Replace with CPC to CPS call - using direct acces to the CPS Store - until thr auth part is clear
+	urlRef, _ := url.Parse(carePlanReference)
+	query := urlRef.Query()
+
+	// ?_include=CarePlan:activity-reference&_include:iterate=Task:focus
+	// query.Add("_include", "CarePlan:activity-reference")
+	// //TODO: This is a tmp solution, but if it's not that temporary, not all FHIR servers support chaining
+	// query.Add("_include:iterate", "Task:focus") //chain the Task.focus from the joined CarePlan.activity-references
+	urlRef.RawQuery = query.Encode()
+
+	// carePlanReference = strings.TrimPrefix(urlRef.Path, "/CarePlan/")
+	startIndex := strings.Index(urlRef.String(), "/CarePlan")
+	if startIndex == -1 {
+		return nil, fmt.Errorf("unable to find CarePlan reference in URL: %s", urlRef.String())
+	}
+
+	localCarePlanRef := urlRef.String()[startIndex:]
+	localCarePlanRef = strings.ReplaceAll(localCarePlanRef, "%3F", "?")
```

Comment:
looks like instead, the URL should be decoded first, then process

Also, I don't think HTTP header values need escaping? I think regular any ASCII character (except newlines) are allowed.